src/v/coproc/script_database.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "coproc/types.h"
+
+#include <absl/container/node_hash_map.h>
+#include <absl/container/node_hash_set.h>
+
+#include <vector>
+
+namespace coproc::wasm {
+/// In memory cache of all coprocessors of all coprocessors
+///
+/// Useful when its necessary to redeploy or reference a shutdown or partitally
+/// shutdown coprocessor for its metadata
+class script_database {
+public:
+    struct script_metadata {
+        iobuf source;
+        std::vector<topic_namespace_policy> inputs;
+    };
+
+    using underlying_t = absl::node_hash_map<script_id, script_metadata>;
+    using inverted_lookup_t = absl::node_hash_map<
+      model::topic_namespace,
+      absl::node_hash_set<script_id>,
+      model::topic_namespace_hash,
+      model::topic_namespace_eq>;
+    using const_iterator = underlying_t::const_iterator;
+    using const_inv_iterator = inverted_lookup_t::const_iterator;
+
+    /// Querying
+    /// ... by scripts
+    bool exists(script_id id) { return _db.find(id) != _db.end(); }
+    const const_iterator find(script_id id) const { return _db.find(id); }
+    const const_iterator cbegin() const { return _db.cbegin(); }
+    const const_iterator cend() const { return _db.cend(); }
```

Comment:
Why return a `const const_iterator` by value?

There are other examples of this below.